NitroxServer/GameLogic/Spawning/EntitySpawner.cs

```diff
@@ -106,5 +110,74 @@ private void SpawnEntities()
                 }
             }
         }
+
+        private bool GetDataFiles(out string lootDistributions, out Dictionary<string, WorldEntityInfo> worldEntityData)
+        {
+            lootDistributions = "";
+            worldEntityData = new Dictionary<string, WorldEntityInfo>();
+            string resourcesPath = "";
+            string gameResourcesPath = Path.Combine(SteamFinder.FindSteamGamePath(264710, "Subnautica"), "Subnautica_Data/resources.assets");
+            AssetsFile resourcesFile;
+            if (File.Exists(gameResourcesPath))
```

Comment:
Not a fan of `File.Exists` before opening a file. It's safer to just open the file (in `new FileStream`) and deal with the exceptions appropriately. That way the code is safe against race conditions with someone/something else accessing the file.

Reply:
That depends. If this is an area where the exception could be raised many times, then it is better to make sure that you're not going to run into that edge case repeatedly. Generating and reacting to an exception is a heavy-handed approach that actually lead to performance problems in some edge cases.

Generating exceptions in .NET can be a pretty hefty task - especially if you have a very deep call stack. Avoiding exceptions when possible (and appropriate) is generally the safer approach.

Reply:
Sure, if this code is in a critical path, then it's better to check `.Exists` first. Even then, the actual opening of the file should also be wrapped in an exception handler and dealt with appropriate on errors. As far as I'm aware, it's possible to lock a file for exclusive access, meaning that no other application can even read it.

Reply:
You're absolutely right that the file could be locked. If we controlled the environment, then we could implement an OS level semaphore that we use to synchronize access to that file.

Now that it is being mentioned - don't we have complete access control to this resource? If so, then I'd say roll a semaphore to address that concern if it is a reality.

That being said - "file exists" is still a different scenario from "file is locked" and should be handled in isolation. In the absence of access control, then handling the exception is the only recourse. Even then, I would expect to see some retry logic in such cases rather than a fail-fast - particularly in this context. 

However; I really don't like the idea of using exceptions in place of control statements. An exception flow should log the exception at a minimum and try to rectify the current exception at most. It would give me heartburn to see an exception being used as an indication to "try something else" when there are alternatives that cleanly express the intent.

Reply:
Agreed, this piece of code is the worst place to use this technique, because `.Exists` is genuinely used for providing alternatives.

However, what I don't like is the unchecked file access. I've been involved in tons of debugging/discussion sessions with other studens, wondering why `file.Exists` sometimes returned `true` while the opening of the file crashed with a `FileNotFound` exception. They knew the file was being touched by another program, but didn't "understand" the programs were executing in parallel. Since then I always try to advice to wrap file IO in a `try-catch` block; rather deal with an edgecase appropriately than to crash the program. And if the file is really expected to be there (for instance, if it's part of the download/installation; not something optional), I usually omit the `.Exists` check entirely.

Again, I should've simply mentioned to wrap the `new FileStream` part in a `try-catch` block rather than comment on this block of alternative-checking. This is really an edgecase since no other program, or thread in this mod should touch the file.

Reply:
Imo, best and cleanest idea is to make a method that checks for file existence, then checks if it can read it. Too much to repeat and would rather not code something knowing it's going to throw exceptions